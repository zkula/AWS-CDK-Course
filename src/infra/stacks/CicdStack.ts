import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { CodePipeline, CodePipelineSource, ShellStep } from 'aws-cdk-lib/pipelines';
import { Construct } from 'constructs';

export class CicdStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new CodePipeline(this, 'AwesomePipeline', {
      pipelineName: 'AwesomePipeline',
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.gitHub('zkula/AWS-CDK-Course', 'main'),
        commands: ['npm ci', 'npx cdk synth'],
      }),
    });
  }
}
