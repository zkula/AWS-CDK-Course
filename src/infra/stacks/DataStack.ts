import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, Table as DynamoDBTable, ITable } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../Utils';
import { Bucket, HttpMethods, IBucket, ObjectOwnership } from 'aws-cdk-lib/aws-s3';

interface DataStackProps extends StackProps {
  stageName: string;
}
export class DataStack extends Stack {
  public readonly spacesTable: ITable;
  public readonly photosBucket: IBucket;

  constructor(scope: Construct, id: string, props: DataStackProps) {
    super(scope, id, props);

    const suffix = getSuffixFromStack(this);

    this.photosBucket = new Bucket(this, 'SpaceFinderPhotos', {
      bucketName: `space-finder-photos-${suffix}`,
      cors: [
        {
          allowedMethods: [HttpMethods.HEAD, HttpMethods.GET, HttpMethods.PUT],
          allowedOrigins: ['*'],
          allowedHeaders: ['*'],
        },
      ],
      // accessControl: BucketAccessControl.PUBLIC_READ,
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
      blockPublicAccess: {
        blockPublicAcls: false,
        blockPublicPolicy: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
      },
    });
    new CfnOutput(this, 'SpaceFinderPhotosBucketName', {
      value: this.photosBucket.bucketName,
    });

    this.spacesTable = new DynamoDBTable(this, 'SpacesTable', {
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING,
      },
      tableName: `SpaceTable-${suffix}`,
    });
  }
}
