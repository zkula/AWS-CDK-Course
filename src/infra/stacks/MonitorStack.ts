import { CfnOutput, Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../Utils';
import { Alarm, Metric, Unit } from 'aws-cdk-lib/aws-cloudwatch';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { EmailSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';

interface MonitorStackProps extends StackProps {
  stageName: string;
}
export class MonitorStack extends Stack {
  constructor(scope: Construct, id: string, props: MonitorStackProps) {
    super(scope, id, props);

    const suffix = getSuffixFromStack(this);

    const alarmTopic = new Topic(this, 'AlarmTopic', {
      displayName: 'AlarmTopic',
      topicName: 'AlarmTopic',
    });

    alarmTopic.addSubscription(new EmailSubscription('zachkula@hotmail.com'));

    const spacesApi400Alarm = new Alarm(this, 'spacesApi4xxAlarm', {
      metric: new Metric({
        metricName: '4XXError',
        namespace: 'AWS/ApiGateway',
        period: Duration.minutes(1),
        statistic: 'Sum',
        unit: Unit.COUNT,
        dimensionsMap: {
          ApiName: 'SpacesApi',
        },
      }),
      evaluationPeriods: 1,
      threshold: 5,
      alarmName: 'SpacesApi4xxAlarm',
    });

    const topicAction = new SnsAction(alarmTopic);
    spacesApi400Alarm.addAlarmAction(topicAction);
    // spacesApi400Alarm.addOkAction(topicAction);
  }
}
