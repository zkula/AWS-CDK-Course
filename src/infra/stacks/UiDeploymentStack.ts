import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../Utils';
import { join } from 'path';
import { existsSync } from 'fs';
import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
import { Distribution, OriginAccessIdentity } from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';

interface UiDeploymentStackProps extends StackProps {
  stageName: string;
}
export class UiDeploymentStack extends Stack {
  constructor(scope: Construct, id: string, props: UiDeploymentStackProps) {
    super(scope, id, props);

    const suffix = getSuffixFromStack(this);

    //Create new Bucket named `space-finder-frontend-${suffix}`
    const deploymentBucket = new Bucket(this, 'uiDeploymentBucket', {
      bucketName: `space-finder-frontend-${suffix}`,
    });

    const uiDir = join(__dirname, '..', '..', '..', '..', 'space-finder-frontend', 'dist');

    if (!existsSync(uiDir)) {
      console.warn('Ui dir not found: ' + uiDir);
      return;
    }

    //Deploy specified contents to bucket
    new BucketDeployment(this, 'SpacesFinderDeployment', {
      destinationBucket: deploymentBucket,
      sources: [Source.asset(uiDir)],
    });

    const originIdentity = new OriginAccessIdentity(this, 'OriginAccessIdentity');
    deploymentBucket.grantRead(originIdentity);

    const distribution = new Distribution(this, 'SpacesFinderDistribution', {
      defaultRootObject: 'index.html',
      defaultBehavior: {
        origin: new S3Origin(deploymentBucket, {
          originAccessIdentity: originIdentity,
        }),
      },
    });

    new CfnOutput(this, 'SpaceFinderUrl', {
      value: distribution.distributionDomainName,
    });
  }
}
