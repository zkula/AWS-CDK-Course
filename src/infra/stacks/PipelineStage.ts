import { Stage, StageProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { LambdaStack } from './LambdaStack';
import { DataStack } from './DataStack';
import { AuthStack } from './AuthStack';
import { ApiStack } from './ApiStack';
import { UiDeploymentStack } from './UiDeploymentStack';
import { MonitorStack } from './MonitorStack';

export class PipelineStage extends Stage {
  constructor(scope: Construct, id: string, props: StageProps) {
    super(scope, id, props);

    const dataStack = new DataStack(this, 'DataStack', {
      stageName: props.stageName,
    });

    const lambdaStack = new LambdaStack(this, 'LambdaStack', {
      spacesTable: dataStack.spacesTable,
      stageName: props.stageName,
    });

    const authStack = new AuthStack(this, 'AuthStack', {
      photosBucket: dataStack.photosBucket,
      stageName: props.stageName,
    });

    new ApiStack(this, 'ApiStack', {
      spacesLambdaIntegration: lambdaStack.spacesLambdaIntegration,
      userPool: authStack.userPool,
      stageName: props.stageName,
    });

    new UiDeploymentStack(this, 'UiDeploymentStack', {
      stageName: props.stageName,
    });

    new MonitorStack(this, 'MonitorStack', {
      stageName: props.stageName,
    });
  }
}
