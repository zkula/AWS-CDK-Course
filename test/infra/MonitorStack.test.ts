import { App } from 'aws-cdk-lib';
import { MonitorStack } from '../../src/infra/stacks/MonitorStack';
import { Capture, Match, Template } from 'aws-cdk-lib/assertions';

describe('Monitor Stack Tests', () => {
  let monitorStackTemplate: Template;
  beforeAll(() => {
    const testApp = new App({
      outdir: 'cdk.out',
    });

    const monitorStack = new MonitorStack(testApp, 'MonitorStack');
    monitorStackTemplate = Template.fromStack(monitorStack);
  });

  test('Email subscription properties', () => {
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Subscription', {
      Endpoint: 'zachkula@hotmail.com',
      Protocol: 'email',
    });
  });

  test('Sns topic properties', () => {
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Topic', {
      DisplayName: 'AlarmTopic',
      TopicName: 'AlarmTopic',
    });
  });

  test('Sns subscription properties - with matchers', () => {
    monitorStackTemplate.hasResourceProperties(
      'AWS::SNS::Subscription',
      Match.objectEquals({
        Endpoint: 'zachkula@hotmail.com',
        Protocol: 'email',
        TopicArn: {
          Ref: Match.stringLikeRegexp('AlarmTopic'),
        },
      })
    );
  });

  test('Alarm Actions', () => {
    const alarmActionsCapture = new Capture();
    monitorStackTemplate.hasResourceProperties('AWS::CloudWatch::Alarm', {
      AlarmActions: alarmActionsCapture,
    });

    expect(alarmActionsCapture.asArray()).toEqual([
      {
        Ref: expect.stringMatching(/^AlarmTopic/),
      },
    ]);
  });

  test('CloudWatch Alarm snapshot', () => {
    const alarm = monitorStackTemplate.findResources('AWS::CloudWatch::Alarm');
    expect(alarm).toMatchSnapshot();
  });

  test('SnsTopic snapshot', () => {
    const snsTopic = monitorStackTemplate.findResources('AWS::SNS::Topic');
    expect(snsTopic).toMatchSnapshot();
  });
});
