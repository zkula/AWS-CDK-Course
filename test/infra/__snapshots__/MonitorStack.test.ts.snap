// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monitor Stack Tests CloudWatch Alarm snapshot 1`] = `
{
  "spacesApi4xxAlarmA15E009C": {
    "Properties": {
      "AlarmActions": [
        {
          "Ref": "AlarmTopicD01E77F9",
        },
      ],
      "AlarmName": "SpacesApi4xxAlarm",
      "ComparisonOperator": "GreaterThanOrEqualToThreshold",
      "Dimensions": [
        {
          "Name": "ApiName",
          "Value": "SpacesApi",
        },
      ],
      "EvaluationPeriods": 1,
      "MetricName": "4XXError",
      "Namespace": "AWS/ApiGateway",
      "Period": 60,
      "Statistic": "Sum",
      "Threshold": 5,
      "Unit": "Count",
    },
    "Type": "AWS::CloudWatch::Alarm",
  },
}
`;

exports[`Monitor Stack Tests SnsTopic snapshot 1`] = `
{
  "AlarmTopicD01E77F9": {
    "Properties": {
      "DisplayName": "AlarmTopic",
      "TopicName": "AlarmTopic",
    },
    "Type": "AWS::SNS::Topic",
  },
}
`;
